app/routes.py

```diff
@@ -1,2 +1,326 @@
+    query_sort_direction = request.args.get("sort")
+    goals_response = []
+
+    if query_sort_direction == "asc":
+        goals = Goal.query.order_by(asc(Goal.title))
+    elif query_sort_direction == "desc": 
+        goals = Goal.query.order_by(desc(Goal.title))
+    else:
+        goals = Goal.query.all()
+
+    for goal in goals:
+        goals_response.append(goal.convert_to_json())
+
+    return jsonify(goals_response), 200
+
+
+@goals_bp.route("/<goal_id>", methods=["GET"], strict_slashes=False)
+def get_single_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    return make_response({"goal":(saved_goal.convert_to_json())}, 200)
+
+
+@goals_bp.route("/<goal_id>", methods=["PUT"], strict_slashes=False)
+def update_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+    
+    form_data = request.get_json()
+
+    saved_goal.title = form_data["title"]
+
+    db.session.commit()
+
+    return make_response({"goal":(saved_goal.convert_to_json())}, 200)
+
+
+@goals_bp.route("/<goal_id>", methods=["DELETE"], strict_slashes=False)
+def delete_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    db.session.delete(saved_goal)
+    db.session.commit()
+    return {"details": f"Goal {saved_goal.goal_id} \"{saved_goal.title}\" successfully deleted"}, 200
+
+
+@goals_bp.route("/<goal_id>/tasks", methods=["POST"], strict_slashes=False)
+def post_tasks_to_goal(goal_id):
+
+    valid_id_or_400(goal_id)
+
+    saved_goal = Goal.query.get_or_404(goal_id)
+
+    # again, request.get_json() gets the JSON body from the HTTP request sent by the client and formats it as a Python dictionary
+    # So form_data is a dictionary that will show all the Goal attributes for the goal (with ID goal_id) that we want to assign tasks to 
+    form_data = request.get_json()
+
+    # The tasks attribute in this request is a list of whatever tasks the client is requesting to post to this goal
+    # Here we're just taking that list out of IDs and sticking it in task_ids
+    task_ids = form_data["task_ids"]
+
+    # looping through the list of task ids that the client is requesting to post to this specific goal  
+    for each_task_id in task_ids:
+        # using each task_id to query a Task object that needs to be updated with its corresponding match_goal_id foreign key
+        updated_task = Task.query.get_or_404(each_task_id)
+        # grabbing that goal_id from the goal queried above, and assigning that value to the match_goal_id foreign key of each updated task
+        # then continuing to loop through until each task in the post request has been updated 
+        updated_task.match_goal_id = saved_goal.goal_id
+
+    db.session.commit()
+
+    return make_response({"id": saved_goal.goal_id, "task_ids": task_ids}, 200)
```

Comment:
This change would have made the above-mentioned bug more obvious. This would also allow the response to include ALL tasks associated with the goal, not just the new ones.
This isn't super intuitive so I'd definitely be down to talk through this more in our next 1:1

```suggestion
    full_task_ids = []
    for task in goal.tasks:
        full_task_ids.append(task.task_id)
        
    return make_response({"id": goal.goal_id, "task_ids": full_task_ids}, 200)
```

Reply:
^ this was the comment I meant!

Reply:
Yup! You're right. I didn't actually test out this code and it wouldn't quite work as is